pkg/pillar/cmd/volumemgr/blob.go

```diff
@@ -104,6 +104,52 @@ func downloadBlob(ctx *volumemgrContext, objType string, sv SignatureVerifier, b
 	return changed
 }
 
+//AddRefToBlobStatus adds the refObject as an reference to the blobs in the given blob list.
+func AddRefToBlobStatus(ctx *volumemgrContext, refObjectID string, blobStatus ...*types.BlobStatus) {
+	updatedBlobs := make([]*types.BlobStatus, 0)
+	for _, blob := range blobStatus {
+		if blob.RefObjects == nil {
+			blob.RefObjects = make(map[string]bool)
+		}
+		if _, present := blob.RefObjects[refObjectID]; !present {
+			log.Infof("MaybeAddContentTreeRefToBlob: Adding RefObject %s as reference to Blob %s",
+				refObjectID, blob.Sha256)
+			blob.RefObjects[refObjectID] = true
+			updatedBlobs = append(updatedBlobs, blob)
+		}
+	}
+	if len(updatedBlobs) > 0 {
+		publishBlobStatus(ctx, updatedBlobs...)
+	}
+}
+
+//RemoveRefFromBlobStatus removes the reference from the blobs in the given blob list.
+func RemoveRefFromBlobStatus(ctx *volumemgrContext, refObjectID string, blobStatus ...*types.BlobStatus) error {
+	for _, blob := range blobStatus {
+		if blob.RefObjects == nil {
+			log.Fatalf("RemoveRefFromBlobStatus: Attempting to remove reference object %s from empty an "+
+				"RefObjects of Blob %s.",
+				refObjectID, blob.Sha256)
+		}
+		if _, present := blob.RefObjects[refObjectID]; present {
+			log.Infof("RemoveRefFromBlobStatus: Removing RefObjects %s as reference to Blob %s",
+				refObjectID, blob.Sha256)
+			delete(blob.RefObjects, refObjectID)
+			if len(blob.RefObjects) == 0 {
+				unpublishBlobStatus(ctx, blob)
+			} else {
+				publishBlobStatus(ctx, blob)
+			}
+		} else {
+			err := fmt.Errorf("RemoveRefFromBlobStatus: Attempting to remove refObject %s which is not present in "+
```

Comment:
If we are going to use this complexity this MUST be a log.Fatal.
But I don't understand how to track where you have added refs so that only 1) only remove the refs you added from object X to a blob and 2) ensure that when object X goes away you drop all of the references on the blob.

We've done that successfully using a RefCount + a HasFooRef boolean for all existing objects. And that has log.Fatal to ensure we don't try to drop references we do not hold. So why do we need something more complex yet less  strict?

Reply:
In this case I suspect you don't need a HasFooRef boolean if you treat the CTStatus Blobs array of pointers as "I have a reference count on this blob". 
But to avoid code evolution from accidentally breaking such an invariant it makes sense to structure/name the functions differently so that the setting/clearing of Blobs is in the same function as the increment/decrement of the reference counts.
Something like
   AddBlobsToContentTree(ctx, status *ContentTreeStatus, blobStatus ...*types.BlobStatus)
and
   RemoveAllBlobsFromContentTree(ctx, status *ContentTreeStatus)
The RefCount in the blobs and the Blobs array must not be modified outside those functions.
(I don't know if you have a case when you remove some some blobs from the CTStatus; if so a third function would make sense.)

Note that if you ever hit a case where a CT includes the same sha aka blob in multiple places in the tree your current approach will fail. I don't think anything prevents folks from constructing such OCI containers.

Reply:
Added changes as per our discussion. Tracking Blob references by a refCount. Also added `AddBlobsToContentTreeStatus` to handle adding blobs to ContentTree and `RemoveBlobsFromContentTreeStatus` to remove blobs from ContentTree